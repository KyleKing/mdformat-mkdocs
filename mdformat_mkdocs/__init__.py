"""An mdformat plugin for mkdocs."""

__version__ = "1.0.4rc1"

from .plugin import (  # noqa: F401
    POSTPROCESSORS,
    RENDERERS,
    add_cli_options,
    update_mdit,
)
