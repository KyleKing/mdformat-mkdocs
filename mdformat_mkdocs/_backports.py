"""Backport of Python 3.11 StrEnum.

From: https://github.com/clbarnes/backports.strenum/blame/fd9a25dacd7161c172e07e4bd673916f85db2cb5/src/backports/strenum/strenum.py

"""

from __future__ import annotations

from enum import Enum
from typing import Type, TypeVar

_S = TypeVar("_S", bound="StrEnum")


class StrEnum(str, Enum):
    """Enum where members are also (and must be) strings"""

    def __new__(cls: Type[_S], *values: str) -> _S:
        if len(values) > 3:
            raise TypeError("too many arguments for str(): %r" % (values,))
        if len(values) == 1:
            # it must be a string
            if not isinstance(values[0], str):
                raise TypeError("%r is not a string" % (values[0],))
        if len(values) >= 2:
            # check that encoding argument is a string
            if not isinstance(values[1], str):
                raise TypeError("encoding must be a string, not %r" % (values[1],))
        if len(values) == 3:
            # check that errors argument is a string
            if not isinstance(values[2], str):
                raise TypeError("errors must be a string, not %r" % (values[2]))

        value = str(*values)
        member = str.__new__(cls, value)
        member._value_ = value
        return member

    __str__ = str.__str__

    @staticmethod
    def _generate_next_value_(
        name: str,
        start: int,
        count: int,
        last_values: list[str],
    ) -> str:
        """Return the lower-cased version of the member name."""
        return name.lower()
