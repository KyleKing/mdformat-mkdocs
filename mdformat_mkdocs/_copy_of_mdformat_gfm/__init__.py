from .plugin import POSTPROCESSORS as GFM_POSTPROCESSORS
from .plugin import RENDERERS as GFM_RENDERERS
from .plugin import update_mdit as gfm_update_mdit
