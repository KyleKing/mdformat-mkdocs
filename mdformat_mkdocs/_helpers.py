"""General Helpers."""

from __future__ import annotations

import re
from functools import wraps
from typing import Callable

EOL = "\n"
"""Line delimiter."""

MKDOCS_INDENT_COUNT = 4
"""Use 4-spaces for mkdocs."""

FILLER_CHAR = "𝕏"  # noqa: RUF001
"""A spacer that is inserted and then removed to ensure proper word wrap."""


def rstrip_result(func: Callable[..., str]) -> Callable[..., str]:
    """Decorator to `rstrip` the function return."""

    @wraps(func)
    def wrapper(*args, **kwargs) -> str:
        return func(*args, **kwargs).rstrip()

    return wrapper


def separate_indent(line: str) -> tuple[str, str]:
    """Separate leading indent from content. Also used by the test suite."""
    re_indent = re.compile(r"(?P<indent>\s*)(?P<content>[^\s]?.*)")
    match = re_indent.match(line)
    assert match is not None  # for pyright # noqa: S101
    return (match["indent"], match["content"])
