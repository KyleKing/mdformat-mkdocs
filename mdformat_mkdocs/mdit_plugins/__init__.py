"""Plugins."""

from ._content_tabs import CONTENT_TAB_MARKERS, content_tabs_plugin
from ._mkdocs_admon import MKDOCS_ADMON_MARKERS, mkdocs_admon_plugin
