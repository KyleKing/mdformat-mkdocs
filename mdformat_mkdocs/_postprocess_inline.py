"""Postprocess inline."""

from __future__ import annotations

from mdformat.renderer import RenderContext, RenderTreeNode

from ._helpers import FILLER_CHAR, MKDOCS_INDENT_COUNT, rstrip_result

FILLER = FILLER_CHAR * (MKDOCS_INDENT_COUNT - 2)  # `mdformat` default is two spaces
"""A spacer that is inserted and then removed to ensure proper word wrap."""


@rstrip_result
def postprocess_inline(text: str, node: RenderTreeNode, context: RenderContext) -> str:
    """Postprocess inline tokens.

    Fix word wrap for lists to account for the change in indentation.
    We fool word wrap by prefixing an unwrappable dummy string of the same length.
    This prefix needs to be later removed (in `_list_item_renderer`).

    Adapted from: https://github.com/hukkin/mdformat-gfm/blob/cf316a121b6cf35cbff7b0ad6e171f287803f8cb/src/mdformat_gfm/plugin.py#L86-L111

    """
    if not context.do_wrap:
        return text
    wrap_mode = context.options["mdformat"]["wrap"]
    if (
        not isinstance(wrap_mode, int)  # noqa: PLR0916
        or FILLER_CHAR in text
        or (node.parent and node.parent.type != "paragraph")
        or (
            node.parent
            and node.parent.parent
            and node.parent.parent.type != "list_item"
        )
    ):
        return text

    _counter = -1
    parent = node.parent
    while parent and parent.type == "paragraph":
        parent = parent.parent
        _counter += 1
    indent_count = max(_counter, 0)

    soft_break = "\x00"
    text = text.lstrip(soft_break).lstrip()
    filler = (FILLER * indent_count)[:-1] if indent_count else ""
    newline_filler = filler + FILLER if indent_count else FILLER[:-1]
    if len(text) > wrap_mode:
        indent_length = MKDOCS_INDENT_COUNT * indent_count
        wrapped_length = -123
        words: list[str] = []
        for word in text.split(soft_break):
            next_length = wrapped_length + len(word)
            if not words:
                words = [filler, word]
                wrapped_length = indent_length + len(word)
            elif next_length > wrap_mode:
                words += [word, newline_filler]
                wrapped_length = indent_length + len(word)
            else:
                words.append(word)
                wrapped_length = next_length + 1
        return soft_break.join(_w for _w in words if _w)
    return f"{filler}{soft_break}{text}" if filler else text
