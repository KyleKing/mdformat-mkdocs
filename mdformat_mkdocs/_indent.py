from __future__ import annotations

import re
from contextlib import suppress
from enum import Enum
from functools import partial, reduce
from typing import Callable, NamedTuple

from mdformat.renderer import RenderContext, RenderTreeNode
from more_itertools import zip_equal

from .mdit_plugins import CONTENT_TAB_MARKERS, MKDOCS_ADMON_MARKERS

MARKERS = CONTENT_TAB_MARKERS.union(MKDOCS_ADMON_MARKERS)
"""All block type markers."""

RE_LIST_ITEM = re.compile(r"(?P<bullet>[\-*\d.]+)\s+(?P<item>.+)")
"""Match `bullet` and `item` against `content`."""

EOL = "\n"
"""Line delimiter."""

# FILLER_CHAR = "𝕏"  # noqa: RUF003
# """A spacer that is inserted and then removed to ensure proper word wrap."""

MKDOCS_INDENT_COUNT = 4
"""Use 4-spaces for mkdocs."""

DEFAULT_INDENT = " " * MKDOCS_INDENT_COUNT
"""Default indent."""


class Syntax(Enum):
    """Non-standard line types."""

    LIST = "LIST"
    START_MARKED = "START_MARKED"
    EDGE_CODE = "EDGE_CODE"
    HTML = "HTML"

    @classmethod
    def from_content(cls, content: str) -> Syntax | None:
        if RE_LIST_ITEM.fullmatch(content):
            return cls.LIST
        if any(content.startswith(f"{marker} ") for marker in MARKERS):
            return cls.START_MARKED
        if content.startswith("```"):
            return cls.EDGE_CODE
        if content.startswith("<"):
            # TODO: Figure out how to handle "markdown in HTML" (like figcaption)
            # PLANNED: for a paragraph - identify if it starts and ends with <>,
            #  then treat as a concurrent block with unmodified inner indentation
            return cls.HTML
        return None


class ParsedLine(NamedTuple):
    """Parsed Line of text."""

    line_num: int
    indent: str
    content: str
    syntax: Syntax | None


class LineResult(NamedTuple):
    """Parsed Line of text."""

    parsed: ParsedLine
    parents: list[ParsedLine]
    prev_list_peers: list[ParsedLine]  # Only applicable for lists


def separate_indent(line: str) -> tuple[str, str]:
    """Separate leading indent from content. Also used by the test suite."""
    re_indent = re.compile(r"(?P<indent>\s*)(?P<content>[^\s]?.*)")
    match = re_indent.match(line)
    assert match is not None  # for pylint # noqa: S101
    return (match["indent"], match["content"])


def is_parent_line(prev_line: LineResult, parsed: ParsedLine) -> bool:
    return bool(prev_line.parsed.content) and len(parsed.indent) > len(
        prev_line.parsed.indent,
    )


def is_peer_list_line(prev_line: LineResult, parsed: ParsedLine) -> bool:
    return (
        len(parsed.indent) == len(prev_line.parsed.indent)
        and prev_line.parsed.syntax == Syntax.LIST
    )


def acc_parsed_lines(acc: list[LineResult], arg: tuple[int, str]) -> list[LineResult]:
    line_num, content = arg
    indent, content = separate_indent(content)
    syntax = Syntax.from_content(content)
    parsed = ParsedLine(
        line_num=line_num,
        indent=indent,
        content=content,
        syntax=syntax,
    )

    parent_idx = 0
    parents = []
    with suppress(StopIteration):
        parent_idx, parent = next(
            (idx, line)
            for idx, line in enumerate(acc[::-1])
            if is_parent_line(line, parsed)
        )
        parents = [*parent.parents, parent.parsed]
    prev_list_peers = []
    if parsed.syntax == Syntax.LIST:
        prev_list_peers += [
            line.parsed
            for line in acc[parent_idx:][::-1]
            if is_peer_list_line(line, parsed)
        ]

    result = LineResult(parsed=parsed, parents=parents, prev_list_peers=prev_list_peers)
    return [*acc, result]


def get_inner_indent(block_indent: str, line_indent: str) -> str:
    return "".join(line_indent[len(block_indent) :])


class BlockIndent(NamedTuple):
    """Track the parsed code block indentation."""

    raw_indent: str
    indent_depth: int


def acc_code_block_indents(
    acc: list[BlockIndent | None],
    line: LineResult,
) -> list[BlockIndent | None]:
    last = (acc or [None])[-1]
    result = last
    if line.parsed.syntax == Syntax.EDGE_CODE:
        result = (
            None  # On second edge, stop tracking
            if last
            # On first edge, start tracking a code block
            else BlockIndent(
                raw_indent=line.parsed.indent,
                indent_depth=len(line.parents),
            )
        )
    return [*acc, result]


def acc_new_indents(
    acc: list[str],
    arg: tuple[LineResult, BlockIndent | None],
) -> list[str]:
    line, code_block_indent = arg

    result = ""
    if line.parsed.content:
        result = DEFAULT_INDENT * len(line.parents)
        if code_block_indent:
            depth = code_block_indent.indent_depth
            extra_indent = get_inner_indent(
                block_indent=code_block_indent.raw_indent,
                line_indent=line.parsed.indent,
            )
            result = DEFAULT_INDENT * depth + extra_indent

    return [*acc, result]


class ParsedText(NamedTuple):
    """Intermediary result of parsing the text."""

    new_indents: list[str]
    new_contents: list[str]
    # Used only for debugging purposes
    lines: list[dict]
    code_block_indents: list[BlockIndent | None]


def acc_new_contents(
    acc: list[str],
    line: LineResult,
    inc_numbers: bool,
    use_sem_break: bool,
) -> list[str]:
    if use_sem_break:
        raise NotImplementedError("Pending text wrap implementation")

    new_content = line.parsed.content
    if list_match := RE_LIST_ITEM.fullmatch(line.parsed.content):
        new_bullet = "-"
        if list_match["bullet"] not in {"-", "*"}:
            new_bullet = f"{len(line.prev_list_peers) + 1 if inc_numbers else 1}."
        new_content = f'{new_bullet} {list_match["item"]}'

    return [*acc, new_content]


def process_text(text: str, inc_numbers: bool, use_sem_break: bool) -> ParsedText:
    """Post-processor to normalize lists."""
    lines = reduce(acc_parsed_lines, enumerate(text.strip().split(EOL)), [])

    code_block_indents = reduce(acc_code_block_indents, lines, [])
    new_indents = reduce(acc_new_indents, zip_equal(lines, code_block_indents), [])

    new_contents = reduce(
        partial(acc_new_contents, inc_numbers=inc_numbers, use_sem_break=use_sem_break),
        lines,
        [],
    )
    return ParsedText(
        new_indents=new_indents,
        new_contents=new_contents,
        lines=[line._asdict() for line in lines],
        code_block_indents=code_block_indents,
    )


def normalize_list(
    text: str,
    node: RenderTreeNode,
    context: RenderContext,
    check_if_align_semantic_breaks_in_lists: Callable[[], bool],  # Attach with partial
) -> str:
    if node.level > 1:
        # Note: this function is called recursively,
        #   so only process the top-level item
        return text

    # Retrieve user-options
    inc_numbers = bool(context.options["mdformat"].get("number"))

    parsed_text = process_text(
        text=text,
        inc_numbers=inc_numbers,
        use_sem_break=check_if_align_semantic_breaks_in_lists(),
    )

    # PLANNED: Need a flat_map (collapse in more-itertools) to handle semantic indents
    return "".join(
        f"{new_indent}{new_content}{EOL}"
        for new_indent, new_content in zip_equal(
            parsed_text.new_indents,
            parsed_text.new_contents,
        )
    )
