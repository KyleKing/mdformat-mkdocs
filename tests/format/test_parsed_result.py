from pathlib import Path

import pytest
from markdown_it.utils import read_fixture_file

from mdformat_mkdocs._indent import process_text

FIXTURE_PATH = Path(__file__).parent / "fixtures/parsed_result.md"
fixtures = read_fixture_file(FIXTURE_PATH)


@pytest.mark.parametrize(
    ("line", "title", "text", "expected"),
    fixtures,
    ids=[f[1] for f in fixtures],
)
def test_parsed_result(line, title, text, expected, snapshot):
    # TODO: Read these settings from the 'title'
    inc_numbers = False
    use_sem_break = False

    output = process_text(
        text=text,
        eol="\n",
        inc_numbers=inc_numbers,
        use_sem_break=use_sem_break,
    )
    assert output == snapshot
