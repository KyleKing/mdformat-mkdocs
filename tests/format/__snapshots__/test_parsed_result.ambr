# serializer version: 1
# name: test_parsed_result[Combination list]
  ParsedText(
    code_block_indents=list([
      None,
      None,
      None,
    ]),
    lines=list([
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='- item 1',
          indent='',
          line_num=0,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='* item 2',
          indent='    ',
          line_num=1,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='* item 2',
            indent='    ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='1. item 3',
          indent='        ',
          line_num=2,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
    ]),
    new_contents=list([
      '- item 1',
      '- item 2',
      '1. item 3',
    ]),
    new_indents=list([
      '',
      '    ',
      '        ',
    ]),
  )
# ---
# name: test_parsed_result[Corrected Indentation from 3x]
  ParsedText(
    code_block_indents=list([
      None,
      None,
      None,
      None,
    ]),
    lines=list([
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='- item 1',
          indent='',
          line_num=0,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 2',
          indent='   ',
          line_num=1,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 2',
            indent='   ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 3',
          indent='      ',
          line_num=2,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 2',
            indent='   ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 3',
            indent='      ',
            line_num=2,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 4',
          indent='         ',
          line_num=3,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
    ]),
    new_contents=list([
      '- item 1',
      '- item 2',
      '- item 3',
      '- item 4',
    ]),
    new_indents=list([
      '',
      '    ',
      '        ',
      '            ',
    ]),
  )
# ---
# name: test_parsed_result[Dashed list]
  ParsedText(
    code_block_indents=list([
      None,
      None,
    ]),
    lines=list([
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='- item 1',
          indent='',
          line_num=0,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 2',
          indent='    ',
          line_num=1,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
    ]),
    new_contents=list([
      '- item 1',
      '- item 2',
    ]),
    new_indents=list([
      '',
      '    ',
    ]),
  )
# ---
# name: test_parsed_result[Handle Jagged Indents 2x]
  ParsedText(
    code_block_indents=list([
      None,
      None,
      None,
      None,
      None,
      None,
      None,
      None,
    ]),
    lines=list([
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='- item 1',
          indent='',
          line_num=0,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 2',
          indent='  ',
          line_num=1,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 2',
            indent='  ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 3',
          indent='    ',
          line_num=2,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 2',
            indent='  ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 3',
            indent='    ',
            line_num=2,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 4',
          indent='      ',
          line_num=3,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 2',
            indent='  ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 3',
            indent='    ',
            line_num=2,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 4',
            indent='      ',
            line_num=3,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 5',
          indent='        ',
          line_num=4,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 2',
            indent='  ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 6',
          indent='    ',
          line_num=5,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='- item 2',
            indent='  ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='- item 7',
          indent='    ',
          line_num=6,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
          ParsedLine(
            content='- item 6',
            indent='    ',
            line_num=5,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
      }),
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='- item 8',
          indent='',
          line_num=7,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
          ParsedLine(
            content='- item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
      }),
    ]),
    new_contents=list([
      '- item 1',
      '- item 2',
      '- item 3',
      '- item 4',
      '- item 5',
      '- item 6',
      '- item 7',
      '- item 8',
    ]),
    new_indents=list([
      '',
      '    ',
      '        ',
      '            ',
      '                ',
      '        ',
      '        ',
      '',
    ]),
  )
# ---
# name: test_parsed_result[Nested Python Classes. Resolves #13: https://github.com/KyleKing/mdformat-mkdocs/issues/13]
  ParsedText(
    code_block_indents=list([
      None,
      None,
      BlockIndent(
        indent_depth=1,
        raw_indent='    ',
      ),
      BlockIndent(
        indent_depth=1,
        raw_indent='    ',
      ),
      BlockIndent(
        indent_depth=1,
        raw_indent='    ',
      ),
      BlockIndent(
        indent_depth=1,
        raw_indent='    ',
      ),
      BlockIndent(
        indent_depth=1,
        raw_indent='    ',
      ),
      BlockIndent(
        indent_depth=1,
        raw_indent='    ',
      ),
      None,
    ]),
    lines=list([
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='1. Add a serializer class',
          indent='',
          line_num=0,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='',
          indent='',
          line_num=1,
          syntax=None,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. Add a serializer class',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='```python',
          indent='    ',
          line_num=2,
          syntax=<Syntax.EDGE_CODE: 'EDGE_CODE'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. Add a serializer class',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='class RecurringEventSerializer(serializers.ModelSerializer):  # (1)!',
          indent='    ',
          line_num=3,
          syntax=None,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. Add a serializer class',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='class RecurringEventSerializer(serializers.ModelSerializer):  # (1)!',
            indent='    ',
            line_num=3,
            syntax=None,
          ),
        ]),
        'parsed': ParsedLine(
          content='"""Used to retrieve recurring_event info"""',
          indent='        ',
          line_num=4,
          syntax=None,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='',
          indent='',
          line_num=5,
          syntax=None,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. Add a serializer class',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='class RecurringEventSerializer(serializers.ModelSerializer):  # (1)!',
            indent='    ',
            line_num=3,
            syntax=None,
          ),
        ]),
        'parsed': ParsedLine(
          content='class Meta:',
          indent='        ',
          line_num=6,
          syntax=None,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. Add a serializer class',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='class RecurringEventSerializer(serializers.ModelSerializer):  # (1)!',
            indent='    ',
            line_num=3,
            syntax=None,
          ),
          ParsedLine(
            content='class Meta:',
            indent='        ',
            line_num=6,
            syntax=None,
          ),
        ]),
        'parsed': ParsedLine(
          content='model = RecurringEvent  # (2)!',
          indent='            ',
          line_num=7,
          syntax=None,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. Add a serializer class',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='```',
          indent='    ',
          line_num=8,
          syntax=<Syntax.EDGE_CODE: 'EDGE_CODE'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
    ]),
    new_contents=list([
      '1. Add a serializer class',
      '',
      '```python',
      'class RecurringEventSerializer(serializers.ModelSerializer):  # (1)!',
      '"""Used to retrieve recurring_event info"""',
      '',
      'class Meta:',
      'model = RecurringEvent  # (2)!',
      '```',
    ]),
    new_indents=list([
      '',
      '',
      '    ',
      '    ',
      '        ',
      '',
      '        ',
      '            ',
      '    ',
    ]),
  )
# ---
# name: test_parsed_result[Numbered list]
  ParsedText(
    code_block_indents=list([
      None,
      None,
      None,
      None,
      None,
      None,
    ]),
    lines=list([
      dict({
        'parents': list([
        ]),
        'parsed': ParsedLine(
          content='1. item 1',
          indent='',
          line_num=0,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='1. item 2-a',
          indent='    ',
          line_num=1,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='1. item 2-a',
            indent='    ',
            line_num=1,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='1. item 3-a',
          indent='        ',
          line_num=2,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='2. item 2-b',
          indent='    ',
          line_num=3,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='2. item 2-b',
            indent='    ',
            line_num=3,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='1. item 3-b',
          indent='        ',
          line_num=4,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
          ParsedLine(
            content='1. item 3-a',
            indent='        ',
            line_num=2,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
      }),
      dict({
        'parents': list([
          ParsedLine(
            content='1. item 1',
            indent='',
            line_num=0,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='2. item 2-b',
            indent='    ',
            line_num=3,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
        'parsed': ParsedLine(
          content='2. item 3-b',
          indent='        ',
          line_num=5,
          syntax=<Syntax.LIST: 'LIST'>,
        ),
        'prev_list_peers': list([
          ParsedLine(
            content='1. item 3-b',
            indent='        ',
            line_num=4,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
          ParsedLine(
            content='1. item 3-a',
            indent='        ',
            line_num=2,
            syntax=<Syntax.LIST: 'LIST'>,
          ),
        ]),
      }),
    ]),
    new_contents=list([
      '1. item 1',
      '1. item 2-a',
      '1. item 3-a',
      '1. item 2-b',
      '1. item 3-b',
      '1. item 3-b',
    ]),
    new_indents=list([
      '',
      '    ',
      '        ',
      '    ',
      '        ',
      '        ',
    ]),
  )
# ---
